pkg/chunk/aws/metrics_autoscaling.go

```diff
@@ -0,0 +1,218 @@
+	case errorRate > 0 &&
+		m.queueLengths[2] > float64(m.queueLengthTarget) &&
+		m.queueLengths[2] > m.queueLengths[1] && m.queueLengths[1] > m.queueLengths[0]:
+		// Growing queue, some errors -> scale up
+		d.scaleUpWrite(current, expected, int64(float64(current.ProvisionedWrite)*m.scaleUpFactor), "metrics queue growing scale-up")
+	}
+	return nil
+}
+
+func (d dynamoTableClient) scaleDownWrite(current, expected *chunk.TableDesc, newWrite int64, msg string) {
+	if newWrite < expected.WriteScale.MinCapacity {
+		newWrite = expected.WriteScale.MinCapacity
+	}
+	// If we're already at or below the requested value, it's not a scale-down.
+	if newWrite >= current.ProvisionedWrite {
+		return
+	}
+	earliest := d.metrics.tableLastUpdated[current.Name].Add(time.Duration(expected.WriteScale.InCooldown) * time.Second)
+	if earliest.After(mtime.Now()) {
+		level.Info(util.Logger).Log("msg", "deferring "+msg, "table", current.Name, "till", earliest)
+		return
+	}
+	// Reject a change that is less than 20% - AWS rate-limits scale-downs so save
+	// our chances until it makes a bigger difference
+	if newWrite > current.ProvisionedWrite*4/5 {
+		level.Info(util.Logger).Log("msg", "rejected de minimis "+msg, "table", current.Name, "current", current.ProvisionedWrite, "proposed", newWrite)
+		return
+	}
+	// Check that the ingesters seem to be doing some work - don't want to scale down
+	// if all our metrics are returning zero, or all the ingesters have crashed, etc
+	totalUsage := 0.0
+	for _, u := range d.metrics.usageRates {
+		totalUsage += u
+	}
+	if totalUsage < minUsageForScaledown {
+		level.Info(util.Logger).Log("msg", "rejected low usage "+msg, "table", current.Name, "totalUsage", totalUsage)
+		return
+	}
+
+	level.Info(util.Logger).Log("msg", msg, "table", current.Name, "write", newWrite)
+	expected.ProvisionedWrite = newWrite
+	d.metrics.tableLastUpdated[current.Name] = mtime.Now()
+}
+
+func (d dynamoTableClient) scaleUpWrite(current, expected *chunk.TableDesc, newWrite int64, msg string) {
+	if newWrite > expected.WriteScale.MaxCapacity {
+		newWrite = expected.WriteScale.MaxCapacity
+	}
+	earliest := d.metrics.tableLastUpdated[current.Name].Add(time.Duration(expected.WriteScale.OutCooldown) * time.Second)
+	if earliest.After(mtime.Now()) {
+		level.Info(util.Logger).Log("msg", "deferring "+msg, "table", current.Name, "till", earliest)
+		return
+	}
+	if newWrite > current.ProvisionedWrite {
+		level.Info(util.Logger).Log("msg", msg, "table", current.Name, "write", newWrite)
+		expected.ProvisionedWrite = newWrite
+		d.metrics.tableLastUpdated[current.Name] = mtime.Now()
+	}
+}
+
+func newMetrics(cfg DynamoDBConfig) (*metricsData, error) {
+	var promAPI promV1.API
+	if cfg.MetricsURL != "" {
+		client, err := promApi.NewClient(promApi.Config{Address: cfg.MetricsURL})
+		if err != nil {
+			return nil, err
+		}
+		promAPI = promV1.NewAPI(client)
```

Comment:
If I read this right, this will return a non-functioning but non-nil `metricsData` if the addr is empty?  That seems odd to me; I'd expect either an error or a fully-functioning`metricsData` - perhaps this logic should be pushed up to the called?